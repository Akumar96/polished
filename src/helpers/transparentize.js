// @flow
import { isRgb, isRgba, isHsl, isHsla, isHex } from '../internalHelpers/_isValidColor'
import hexToRgb from '../internalHelpers/_hexToRgb'
import hslToRgb from '../internalHelpers/_hslToRgb'

const getNewTransparency = (currentTransparency: number, percentage: number) : number => (
  currentTransparency * (1 - percentage)
)

const getColorValuesInBrackets = (color: string) : string => (
  color.slice(color.indexOf('(') + 1, color.indexOf(')'))
)

const getValuesTillLastComma = (color: string) : string => (
  color.slice(0, color.lastIndexOf(','))
)

const transparentizeRgb = (rgb: string, percentage: number) : string => (
  `rgba(${getColorValuesInBrackets(rgb)}, ${getNewTransparency(1, percentage)})`
)

const getLastColorValue = (color: string) : number => (
  Number(
    color.slice(
      color.lastIndexOf(',') + 1,
      color.indexOf(')'),
    ),
  )
)

function transparentize(color: string, percentage: number) {
  if (percentage >= 0 && percentage <= 1) {
    if (isHex(color)) {
      return transparentizeRgb(`${hexToRgb(color)})`, percentage)
    } else if (isRgb(color)) {
      return transparentizeRgb(color, percentage)
    } else if (isRgba(color)) {
      const currentTransparency = getLastColorValue(color)
      const rgb = getValuesTillLastComma(getColorValuesInBrackets(color))
      const newTransparency = getNewTransparency(currentTransparency, percentage)

      return `rgba(${rgb}, ${newTransparency})`
    } else if (isHsl(color)) {
      const newTransparency = getNewTransparency(1, percentage)
      const values = getColorValuesInBrackets(color).split(',').map(num => parseInt(num, 10) / 100)
      const rgbValues = hslToRgb(...values)

      return `rgba(${rgbValues}, ${newTransparency})`
    } else if (isHsla(color)) {
      const hslValues = getValuesTillLastComma(getColorValuesInBrackets(color))
      const currentTransparency = getLastColorValue(color)
      const newTransparency = getNewTransparency(currentTransparency, percentage)
      const values = hslValues.split(',').map(num => parseInt(num, 10) / 100)
      const rgbValues = hslToRgb(...values)

      return `rgba(${rgbValues}, ${newTransparency})`
    } else {
      throw new Error('Invalid color')
    }
  } else {
    throw new Error('Invalid percentage, only values from 0 to 1 are accepted')
  }
}

export default transparentize
