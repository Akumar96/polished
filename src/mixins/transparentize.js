// @flow
import { isRgb, isRgba, isHsl, isHsla, isHex } from '../internalHelpers/_isValidColor'
import hslToRgb from '../internalHelpers/_hslToRgb'
import hexToRgb from '../internalHelpers/_hexToRgb'

const getNewTransparency = (currentTransparency: number, percentage: number) : number => (
  currentTransparency * (1 - (percentage / 100))
)

const getColorValuesInBrackets = (color: string) : string => (
  color.slice(color.indexOf('(') + 1, color.indexOf(')'))
)

const getValuesTillLastComma = (color: string) : string => (
  color.slice(0, color.lastIndexOf(','))
)

const transparentizeRgb = (rgb: string, percentage: number) : string => (
  `rgba(${getColorValuesInBrackets(rgb)}, ${getNewTransparency(1, percentage)})`
)

const getLastColorValue = (color: string) : number => (
  Number(
    color.slice(
      color.lastIndexOf(',') + 1,
      color.indexOf(')'),
    ),
  )
)

function transparentize(color: string, percentage : number) {
  if (isHex(color)) {
    return transparentizeRgb(`${hexToRgb(color)})`, percentage)
  } else if (isRgb(color)) {
    return transparentizeRgb(color, percentage)
  } else if (isRgba(color)) {
    const currentTransparency = getLastColorValue(color)
    const rgb = getValuesTillLastComma(getColorValuesInBrackets(color))
    const newTransparency = getNewTransparency(currentTransparency, percentage)

    return `rgba(${rgb}, ${newTransparency})`
  } else if (isHsl(color)) {
    const hslValues = getColorValuesInBrackets(color)
    const rgb = hslToRgb(
      ...(hslValues.split(', ').map(val => parseInt(val, 10))),
    )

    return transparentizeRgb(rgb, percentage)
  } else if (isHsla(color)) {
    const hslValues = getValuesTillLastComma(getColorValuesInBrackets(color))
    const currentTransparency = getLastColorValue(color)
    const rgb = hslToRgb(
      ...(hslValues.split(', ').map(val => parseInt(val, 10))),
    )

    const newTransparency = getNewTransparency(currentTransparency, percentage)
    return `rgba(${rgb}, ${newTransparency})`
  } else {
    throw new Error('Invalid color')
  }
}

export default transparentize
