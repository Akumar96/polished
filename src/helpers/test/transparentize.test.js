// @flow
import transparentize from '../transparentize'

describe('transparentize', () => {
  it('it should correctly convert hex #fff with 0.1 amount transparency', () => {
    expect(transparentize(0.1, '#fff')).toMatchSnapshot()
  })

  it('it should correctly convert rgb(255, 0, 255) with 0.1 amount transparency', () => {
    expect(transparentize(0.1, 'rgb(255, 0, 255)')).toMatchSnapshot()
  })

  it('it should correctly convert rgba(255, 0, 0, 1) with 0.1 amount transparency', () => {
    expect(transparentize(0.1, 'rgba(255, 0, 0, 1)')).toMatchSnapshot()
  })

  it('it should correctly convert rgba(255, 0, 0, .5) with 0.3 amount transparency', () => {
    expect(transparentize(0.3, 'rgba(255, 0, 0, .5)')).toMatchSnapshot()
  })

  it('it should correctly convert rgba(255, 0, 0, .5) with 0.5 amount transparency', () => {
    expect(transparentize(0.5, 'rgba(255, 0, 0, .5)')).toMatchSnapshot()
  })

  it('it should correctly convert hsl(0, 0%, 100%) with 0.2 amount transparency', () => {
    expect(transparentize(0.2, 'hsl(0, 0%, 100%)')).toMatchSnapshot()
  })

  it('it should correctly convert hsla(0, 0%, 100%, .8) with 0.5 amount transparency', () => {
    expect(transparentize(0.5, 'hsla(0, 0%, 100%, .8)')).toMatchSnapshot()
  })

  it('it should throw an error when enter an invalid color', () => {
    expect(() => { transparentize(0.5, 'not a color') }).toThrow(new Error('Invalid color'))
  })

  it('it should throw an error when enter an invalid percentage', () => {
    expect(() => { transparentize(1.5, 'rgba(255, 255, 255, .5)') })
      .toThrow(new Error('Invalid amount, only values from 0 to 1 are accepted'))
  })
})
