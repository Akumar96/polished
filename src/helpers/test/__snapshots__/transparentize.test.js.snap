exports[`transparentize it should correctly convert hex #fff with 0.1 amount transparency 1`] = `"rgba(255, 255, 255, 0.1)"`;

exports[`transparentize it should correctly convert hsl(0, 0%, 100%) with 0.2 amount transparency 1`] = `"rgba(255,255,255, 0.2)"`;

exports[`transparentize it should correctly convert hsla(0, 0%, 100%, .8) with 0.5 amount transparency 1`] = `"rgba(255,255,255, 0.5)"`;

exports[`transparentize it should correctly convert rgb(255, 0, 255) with 0.1 amount transparency 1`] = `"rgba(255, 0, 255, 0.1)"`;

exports[`transparentize it should correctly convert rgba(255, 0, 0, .5) with 0.3 amount transparency 1`] = `"rgba(255, 0, 0, 0.3)"`;

exports[`transparentize it should correctly convert rgba(255, 0, 0, .5) with 0.5 amount transparency 1`] = `"rgba(255, 0, 0, 0.5)"`;

exports[`transparentize it should correctly convert rgba(255, 0, 0, 1) with 0.1 amount transparency 1`] = `"rgba(255, 0, 0, 0.1)"`;
